app/get_proc_arg.go

```diff
@@ -0,0 +1,221 @@
+	DefaultFlexvolumeDir = "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
+
+	DetectPodMaxPolls      = 120
+	DetectFlexvolumeName   = "discover-flexvolume-dir"
+	DetectRootDirName      = "discover-kubelet-root-dir"
+	DetectKubeletArgScript = `
+    find_kubelet_proc() {
+      for proc in $(find /proc -maxdepth 1 -type d 2>/dev/null); do
+        if [ ! -f $proc/cmdline ]; then
+          continue
+        fi
+        if [[ "$(cat $proc/cmdline | tr '\000' '\n' | head -n1 | tr '/' '\n' | tail -n1)" == "kubelet" ]]; then
+          echo $proc
+        fi
+      done
+    }
+    get_kubelet_arg_dir() {
+      proc=$(find_kubelet_proc)
+      if [ "$proc" != "" ]; then
+        path=$(cat $proc/cmdline | tr '\000' '\n' | grep %s | tr '=' '\n' | tail -n1)
+        if [ "$path" == "" ]; then
+          echo %s
+        else
+          echo $path
+		fi
+      fi
+    }
+	get_kubelet_arg_dir
+	`
+	DetectK3SDataDirScript = `
+    find_k3s_proc() {
+      for proc in $(find /proc -maxdepth 1 -type d 2>/dev/null); do
+        if [ ! -f $proc/cmdline ]; then
+          continue
+        fi
+        if [[ $(cat $proc/cmdline | tr '\000' '\n' | sed -n '1p') == "k3s" ]]; then
+          proc_name=$(cat $proc/cmdline | tr '\000' '\n' | sed -n '2p')
+          if [[ "$proc_name"  == "server" || "$proc_name" == "agent" ]]; then
+            echo $proc
+          fi
+        fi
+      done
+    }
+    get_k3s_data_dir() {
+      proc=$(find_k3s_proc)
+      if [ "$proc" != "" ]; then
+        path=$(cat $proc//cmdline | tr '\000' '\n' |  grep 'data-dir' -A 1 | tail -n1)
+        if [ "$path" == "" ]; then
+          path=$(cat $proc//cmdline | tr '\000' '\n' |  grep '\^-d$' -A 1 | tail -n1)
+        fi
+        if [ "$path" == "" ]; then
+          echo '/var/lib/rancher/k3s'
+        else
+          echo $path
+        fi
+      fi
+    }
+    get_k3s_data_dir
+	`
+)
+
+func getProcArg(kubeClient *clientset.Clientset, managerImage, name string) (string, error) {
+	switch name {
+	case ArgFlexvolumePluginDir:
+		dir, err := discoverFlexvolumeDir(kubeClient, managerImage)
+		if err != nil {
+			logrus.Errorf("Failed to get arg volume-plugin-dir, err: %v", err)
+			return "", err
+		}
+		return dir, nil
+	case ArgRootDir:
+		dir, err := discoverRootDir(kubeClient, managerImage)
+		if err != nil {
+			logrus.Errorf("Failed to get arg root-dir, err: %v", err)
+			return "", err
+		}
+		return dir, nil
+	}
+	return "", fmt.Errorf("Getting arg %v is not supported", name)
+}
+
+func discoverFlexvolumeDir(kubeClient *clientset.Clientset, managerImage string) (string, error) {
+	detectFlexVolumeScript := fmt.Sprintf(DetectKubeletArgScript, ArgFlexvolumePluginDir, DefaultFlexvolumeDir)
+	dir, err := discoverProcArg(kubeClient, managerImage, DetectFlexvolumeName, detectFlexVolumeScript)
+	if err != nil {
+		logrus.Errorf("Failed to find arg volume-plugin-dir in proc kubelet, err: %v", err)
+		return "", err
+	}
+	if dir == "" {
+		dir = DefaultFlexvolumeDir
```

Comment:
Does this mean we cannot find kubelet process? (since the default was already set inside the script) so we should error out here?

Same below.

Reply:
this means we cannot find both kubelet and k3s. Here I just use default value, but we can error out instead.